src/store/ducks/opensrp/assignments/index.ts

```diff
@@ -0,0 +1,159 @@
+import SeamlessImmutable from 'seamless-immutable';
+
+/** The reducer name */
+export const reducerName = 'assignments';
+
+/** interface for a Assignment object */
+export interface Assignment {
+  jurisdiction: string;
+  organization: string;
+  plan: string;
+  fromDate: string;
+  toDate: string;
+}
+
+// action interfaces
+
+/** action type for action that adds Assignments to store */
+export const ASSIGNMENTS_FETCHED = 'src/store/ducks/assignments/reducer/ASSIGNMENTS_FETCHED';
+/** action type for REMOVE_TEAMS action */
+export const REMOVE_ASSIGNMENTS = 'src/store/ducks/assignments/reducer/REMOVE_ASSIGNMENTS';
+
+/** interface for Assignments fetched action */
+interface FetchAssignmentsAction extends AnyAction {
+  assignmentsByPlanId: { [key: string]: Assignment[] };
+  type: typeof ASSIGNMENTS_FETCHED;
+}
+
+/** interface for action that removes assignments from store */
+interface RemoveAssignmentsAction extends AnyAction {
+  assignmentsByPlanId: { [key: string]: Assignment[] };
+  type: typeof REMOVE_ASSIGNMENTS;
+}
+
+/** single type for all action types */
+type AssignmentActionTypes = FetchAssignmentsAction | RemoveAssignmentsAction | AnyAction;
+
+/** interface describing the assignmentsByPlanId Store */
+interface AssignmentsByPlanId {
+  [key: string]: Assignment[];
+}
+/** interface for Assignments state in store */
+interface AssignmentsStoreState {
+  assignmentsByPlanId: AssignmentsByPlanId;
+}
+
+// immutable assignments state in dux
+export type ImmutableAssignmentsStoreState = AssignmentsStoreState &
+  SeamlessImmutable.ImmutableObject<AssignmentsStoreState>;
+
+/** initial state for Assignments records in store */
+const initialOrgsStoreState: ImmutableAssignmentsStoreState = SeamlessImmutable({
+  assignmentsByPlanId: {},
+});
+
+/** the Assignment reducer function */
+export default function reducer(state = initialOrgsStoreState, action: AssignmentActionTypes) {
+  switch (action.type) {
+    case ASSIGNMENTS_FETCHED:
+      return SeamlessImmutable({
+        ...state,
+        assignmentsByPlanId: {
+          ...state.assignmentsByPlanId,
+          ...action.assignmentsByPlanId,
+        },
+      });
+    case REMOVE_ASSIGNMENTS:
+      return SeamlessImmutable({
+        ...state,
+        assignmentsByPlanId: action.assignmentsByPlanId,
+      });
+    default:
+      return state;
+  }
+}
+
+/** action to remove assignments form store */
+export const removeAssignmentsAction: RemoveAssignmentsAction = {
+  assignmentsByPlanId: {},
+  type: REMOVE_ASSIGNMENTS,
+};
+
+// action creators
+
+/** creates action to add fetched assignments to store
+ * @param {Assignment []} assignmentsList - array of assignments to be added to store
+ *
+ * @returns {FetchAssignmentsAction} - action with assignments payload that is added to store
+ */
+export const fetchAssignments = (assignmentsList: Assignment[]): FetchAssignmentsAction => {
+  const assignmentsByPlanId: AssignmentsByPlanId = {};
```

Comment:
also, was looking into a way to do default values for JS objects, you can do this with a proxy, but I'm not sure how common this is, although it looks closest to how we'd do it in other languages with hash default values:

```javascript
var arrayDefault = {
  get: function(target, name) {
    return target.hasOwnProperty(name) ? target[name] : [];
  }
};

const assignmentsByPlanId: AssignmenstByPlanId = Proxy({}, arrayDefault)
```

and then below should simplify to

```javascript
for (const assignment of assignmentsList) {
  assignmentsByPlanId[assignment.plan].push(assignment);
}
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

Reply:
Ticket created here: https://github.com/onaio/reveal-frontend/issues/527